### PR TITLE
Make wee-alloc an optional feature.

### PR DESCRIPTION
**Pull request by yuval-k**
Allow building for non-wasm arch by only depending on wee allocator if we are compiling to wasm32.
Compiling to native has the advantages of being able easily write and run unit tests in dependent projects.

I did test this manually to work, but I am by no means a bazel expert.

### CODE REVIEW

**Comment by yuval-k** [comment]
@PiotrSikora given that I didn't really touched any code, how do you want to handle the failed outdated check?
I can update the dependency, but I'm not sure if that's desired.

**Comment by PiotrSikora** [comment]
> @PiotrSikora given that I didn't really touched any code, how do you want to handle the failed outdated check?
> I can update the dependency, but I'm not sure if that's desired.

Please ignore it. Those dependencies that were outdated before. I'll fix it shortly.

**Comment by gbrail** [comment]
FWIW I think that we should merge this or something like it.

I actually wish that the library didn't bring in wee-alloc at all. I don't think that it's the business of a library to set the global allocator on any platform, and WASM targets work just fine without it. The only downside is a few tens of K of generated code, which in any module I've written for Envoy is dwarfed by all the other stuff. Projects that want to shrink the size of their WASM can always include it themselves.

**Comment by PiotrSikora** [comment]
@gbrail the original reason for enabling this was indeed shrining the `.wasm` size, but a lot has changed since, and I agree that `wee_alloc` shouldn't be forced on users.

@yuval-k could you turn it into a crate features, by any chance?

**Comment by yuval-k** [comment]
i will give it a try; but not an expert on bazel machinery. will update soon

**Comment by yuval-k** [comment]
@PiotrSikora I added the feature to the rust side, but I'm not sure how to do it on the bazel side.. do you happen to know that off the top of your head? if not i'll have to dig in to the bazel rust rules...

File: Cargo.toml [hunk 1]
```diff
@@ -11,7 +11,10 @@ edition = "2018"
 [dependencies]
 hashbrown = { version = "0.8", default-features = false, features = ["ahash", "inline-more"] }
 log = "0.4"
-wee_alloc = "0.4"
+wee_alloc = { version = "0.4", optional = true }
+
+[features]
+wee-alloc = ["wee_alloc"]
```
Comment (PiotrSikora) [inline Cargo.toml:17]: Could you move `[features]` before `[dependencies]`?

**Review by PiotrSikora — COMMENTED** [review]
Hmm... It looks that one of my responses got lost.

You need to regenerate `BUILD` files to match updated `Cargo.toml`:
```
cargo install cargo-raze --version 0.3.8
rm -rf bazel/cargo/
cargo generate-lockfile
cargo raze --output=bazel/cargo
mv Cargo.lock bazel/cargo/
```

Regarding Bazel, I couldn't find a way to make `cargo-raze` support features in the primary crate, but I think that we can skip `wee_alloc` there for now.

**Comment by PiotrSikora** [comment]
> Please ignore it. Those dependencies that were outdated before. I'll fix it shortly.

Once #45 is merged, you should merge/rebase to fix outdated check. Sorry for the delay!

**Comment by yuval-k** [comment]
sounds good; will also add the raze comment to the readme

**Comment by yuval-k** [comment]
just tested this;
The problem with allow `malloc` is that it creates infinite recursion when not building for wasm:

```
#151 0x00005555555a6626 in malloc ()
#152 0x00005555555a5b0c in alloc::alloc::alloc ()
#153 0x00005555555a5c71 in <alloc::alloc::Global as core::alloc::AllocRef>::alloc ()
#154 0x00005555555a4743 in alloc::raw_vec::RawVec<T,A>::allocate_in ()
#155 0x00005555555a4a68 in alloc::raw_vec::RawVec<T,A>::with_capacity_in ()
#156 0x00005555555a445a in alloc::raw_vec::RawVec<T>::with_capacity ()
#157 0x00005555555a510a in alloc::vec::Vec<T>::with_capacity ()
#158 0x00005555555a6626 in malloc ()
#159 0x00005555555a5b0c in alloc::alloc::alloc ()
#160 0x00005555555a5c71 in <alloc::alloc::Global as core::alloc::AllocRef>::alloc ()
#161 0x00005555555a4743 in alloc::raw_vec::RawVec<T,A>::allocate_in ()
#162 0x00005555555a4a68 in alloc::raw_vec::RawVec<T,A>::with_capacity_in ()
#163 0x00005555555a445a in alloc::raw_vec::RawVec<T>::with_capacity ()
#164 0x00005555555a510a in alloc::vec::Vec<T>::with_capacity ()
#165 0x00005555555a6626 in malloc ()
#166 0x00005555555a5b0c in alloc::alloc::alloc ()
#167 0x00005555555a5c71 in <alloc::alloc::Global as core::alloc::AllocRef>::alloc ()
#168 0x00005555555a4743 in alloc::raw_vec::RawVec<T,A>::allocate_in ()
#169 0x00005555555a4a68 in alloc::raw_vec::RawVec<T,A>::with_capacity_in ()
#170 0x00005555555a445a in alloc::raw_vec::RawVec<T>::with_capacity ()
#171 0x00005555555a510a in alloc::vec::Vec<T>::with_capacity ()
#172 0x00005555555a6626 in malloc ()
...
```

To fix this I added ` #[cfg(target_arch = "wasm32")] ` to mod allocator

**Comment by PiotrSikora** [comment]
@yuval-k ah, so your issue is not with `wee_alloc` itself after all, but `malloc`? Or both? Could you try https://github.com/proxy-wasm/proxy-wasm-rust-sdk/pull/42?

Note that even if #42 fixes your issue, I'd still like to get merge this PR for adding `wee_alloc` as optional feature.

**Comment by yuval-k** [comment]
it's both; wee_alloc (when doesn't target wasm) brings in `libc` create; so it's easier to not include it than include both

**Comment by PiotrSikora** [comment]
@yuval-k could you merge/rebase on top of master with merged #42 and try to only guard `#[global_allocator]`? Removing `malloc` / `proxy_on_memory_allocate` doesn't seem correct, since it's part of the ABI.

**Comment by yuval-k** [comment]
yes, that works!

**Comment by PiotrSikora** [comment]
One more thing, could you please add `Build (wasm32-wasi with wee-alloc)` and `Clippy (wasm32-wasi with wee-alloc)` with `--features=wee-alloc` to both `stable` and `nightly` jobs in `.github/workflows/rust.yml`?

File: README.md [hunk 2]
```diff
@@ -25,3 +25,14 @@
 
 + [Extending Envoy with WASM and Rust](https://antweiss.com/blog/extending-envoy-with-wasm-and-rust/)
 + [Extending Istio with Rust and WebAssembly](https://blog.red-badger.com/extending-istio-with-rust-and-webassembly)
+
+# Updating dependencies
+
+When updating dependencies, You need to regenerate `BUILD` files to match updated `Cargo.toml`:
```
Comment (PiotrSikora) [inline README.md:31]: Nit: `s/You/you/`

**Review by PiotrSikora — APPROVED** [review]

File: .github/workflows/rust.yml [hunk 3]
```diff
@@ -106,11 +106,21 @@ jobs:
...
+      env:
+        RUSTFLAGS: -D warnings -C link-args=-S
+      run: cargo build --release --all-targets --target=wasm32-wasi --features=wee-alloc
+
     - name: Clippy (wasm32-wasi)
       env:
         RUSTFLAGS: -D warnings -C link-args=-S
       run: cargo clippy --release --all-targets --target=wasm32-wasi
 
+    - name: Clippy (wasm32-wasi with wee-alloc))
```
Comment (PiotrSikora) [inline .github/workflows/rust.yml:119]: Nit: `))` -> `)`

File: .github/workflows/rust.yml [hunk 4]
```diff
@@ -151,11 +161,21 @@ jobs:
...
```
Comment (PiotrSikora) [inline .github/workflows/rust.yml:174]: Nit: `))` -> `)`

**Comment by PiotrSikora** [comment]
Thanks!

**Pull request merged**